src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php

```diff
@@ -168,13 +172,31 @@ public function load(array $configs, ContainerBuilder $container)
         $container->getDefinition('security.authentication.guard_handler')
             ->replaceArgument(2, $this->statelessFirewallKeys);
 
+        // @deprecated since Symfony 5.3
         if ($config['encoders']) {
             $this->createEncoders($config['encoders'], $container);
+            $config['password_hashers'] = $config['encoders'];
+        } else {
+            $container->removeAlias('security.encoder_factory');
+            $container->removeAlias('security.password_encoder');
+            $container->removeDefinition('security.user_password_encoder.generic');
+        }
+
+        if ($config['password_hashers']) {
+            $this->createHashers($config['password_hashers'], $container);
```

Comment:
I see an issue here: if a project migrate their config to use `password_hashers`, the corresponding encoder services are not defined anymore. And this means that any bundle which is not migrated yet to the new API will break as `security.encoder_factory` won't be able to give it an encoder.
As the feature set is the same, I think we should create legacy services as well when using the new config (similar to how the old config creates new services with your assignment above). If you never use the old `security.encoder_factory` API, having these deprecated services defined won't hurt (deprecated services are allowed to use deprecated classes without triggering a compile-time deprecation warning for that in the compiler pass IIRC) and it will keep BC for services using them (with a deprecation warning)

Reply:
btw, this also implies that both the new and the old factories should support both `*Aware` interfaces implemented on User classes, so that both factories can be used for migrated and non-migrated user classes while getting the right hasher/encoder.